src/modules/bulletproofs/rangeproof_impl.h

```diff
@@ -0,0 +1,792 @@
+        if (overflow || secp256k1_scalar_is_zero(&ecmult_data[i].z)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+
+        secp256k1_bulletproof_deserialize_point(&ecmult_data[i].t1, &proof[i][64], 2, 4);
+        secp256k1_bulletproof_deserialize_point(&ecmult_data[i].t2, &proof[i][64], 3, 4);
+
+        secp256k1_bulletproof_update_commit(commit, &ecmult_data[i].t1, &ecmult_data[i].t2);
+        secp256k1_scalar_set_b32(&ecmult_data[i].x, commit, &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&ecmult_data[i].x)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+
+        /* compute exponent offsets */
+        secp256k1_scalar_inverse_var(&ecmult_data[i].yinv, &ecmult_data[i].y);  /* TODO somehow batch this w the inner-product argument inverse */
+        ecmult_data[i].yinvn = ecmult_data[i].yinv;
+        for (j = 0; j < secp256k1_floor_lg(nbits); j++) {
+            secp256k1_scalar_sqr(&ecmult_data[i].yinvn, &ecmult_data[i].yinvn);
+        }
+        secp256k1_scalar_sqr(&ecmult_data[i].zsq, &ecmult_data[i].z);
+        secp256k1_scalar_negate(&ecmult_data[i].negz, &ecmult_data[i].z);
+
+        /* Update commit with remaining data for the inner product proof */
+        secp256k1_sha256_initialize(&sha256);
+        secp256k1_sha256_write(&sha256, commit, 32);
+        secp256k1_sha256_write(&sha256, &proof[i][0], 64);
+        secp256k1_sha256_finalize(&sha256, commit);
+
+        secp256k1_sha256_initialize(&sha256);
+        secp256k1_sha256_write(&sha256, commit, 32);
+        secp256k1_sha256_finalize(&sha256, randomizer61);
+        secp256k1_scalar_set_b32(&ecmult_data[i].randomizer61, randomizer61, &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&ecmult_data[i].randomizer61)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+
+        /* Deserialize everything else */
+        secp256k1_scalar_set_b32(&taux, &proof[i][0], &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&taux)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+        secp256k1_scalar_set_b32(&mu, &proof[i][32], &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&mu)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+        /* A little sketchy, we read t (l(x) . r(x)) off the front of the inner product proof,
+         * which we otherwise treat as a black box */
+        secp256k1_scalar_set_b32(&ecmult_data[i].t, &proof[i][64 + 128 + 1], &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&ecmult_data[i].t)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+
+        /* Verify inner product proof */
+        ecmult_data[i].a = age;
+        ecmult_data[i].s = sge;
+        ecmult_data[i].n = nbits * n_commits;
```

Comment:
Is `n` actually used? Or do we use the fourth argument of `secp256k1_bulletproof_inner_product_verify_impl()` instead`

Reply:
Yeah, it's used in `secp256k1_bulletproof_rangeproof_vfy_callback` in this file. The inner product code doesn't see this but the rangeproof stuff does.